linebot/insight.go

```diff
@@ -0,0 +1,147 @@
+const (
+	InsightTypeMessageDelivery InsightType = "message/delivery"
+	InsightTypeFollowers       InsightType = "followers"
+	InsightTypeDemographic     InsightType = "demographic"
+)
+
+// GetNumberMessagesDeliveryCall type
+type GetNumberMessagesDeliveryCall struct {
+	c   *Client
+	ctx context.Context
+
+	date        string
+	insightType InsightType
+}
+
+// GetNumberMessagesDelivery method
+func (client *Client) GetNumberMessagesDelivery(date string) *GetNumberMessagesDeliveryCall {
+	return &GetNumberMessagesDeliveryCall{
+		c:           client,
+		date:        date,
+		insightType: InsightTypeMessageDelivery,
+	}
+}
+
+// WithContext method
+func (call *GetNumberMessagesDeliveryCall) WithContext(ctx context.Context) *GetNumberMessagesDeliveryCall {
+	call.ctx = ctx
+	return call
+}
+
+// Do method
+func (call *GetNumberMessagesDeliveryCall) Do() (*MessagesNumberDeliveryResponse, error) {
+	endpoint := fmt.Sprintf(APIEndpointInsight, call.insightType)
+	var q url.Values
+	if call.date != "" {
+		q = url.Values{"date": []string{call.date}}
+	}
+	res, err := call.c.get(call.ctx, endpoint, q)
+	if err != nil {
+		return nil, err
+	}
+	defer closeResponse(res)
+	return decodeToMessagesNumberDeliveryResponse(res)
+}
+
+// GetNumberFollowersCall type
+type GetNumberFollowersCall struct {
+	c   *Client
+	ctx context.Context
+
+	date        string
+	insightType InsightType
+}
+
+// GetNumberFollowers method
+func (client *Client) GetNumberFollowers(date string) *GetNumberFollowersCall {
+	return &GetNumberFollowersCall{
+		c:           client,
+		date:        date,
+		insightType: InsightTypeFollowers,
+	}
+}
+
+// WithContext method
+func (call *GetNumberFollowersCall) WithContext(ctx context.Context) *GetNumberFollowersCall {
+	call.ctx = ctx
+	return call
+}
+
+// Do method
+func (call *GetNumberFollowersCall) Do() (*MessagesNumberFollowersResponse, error) {
+	endpoint := fmt.Sprintf(APIEndpointInsight, call.insightType)
+	var q url.Values
+	if call.date != "" {
+		q = url.Values{"date": []string{call.date}}
+	}
+	res, err := call.c.get(call.ctx, endpoint, q)
+	if err != nil {
+		return nil, err
+	}
+	defer closeResponse(res)
+	return decodeToMessagesNumberFollowersResponse(res)
+}
+
+// GetFriendDemographicsCall type
+type GetFriendDemographicsCall struct {
+	c   *Client
+	ctx context.Context
+
+	date        string
+	insightType InsightType
+}
+
+// GetFriendDemographics method
+func (client *Client) GetFriendDemographics() *GetFriendDemographicsCall {
+	return &GetFriendDemographicsCall{
+		c:           client,
+		insightType: InsightTypeDemographic,
+	}
+}
+
+// WithContext method
+func (call *GetFriendDemographicsCall) WithContext(ctx context.Context) *GetFriendDemographicsCall {
+	call.ctx = ctx
+	return call
+}
+
+// Do method
+func (call *GetFriendDemographicsCall) Do() (*MessagesFriendDemographicsResponse, error) {
+	endpoint := fmt.Sprintf(APIEndpointInsight, call.insightType)
+	var q url.Values
+	if call.date != "" {
```

Comment:
Seems there's no parameter `date` for /insight/demographic

Reply:
Fixed it, thank you.